File: src/symnmf.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "matrix.c"
#include "matrix.h"

#define MAX_ROW_LEN 1024 /* Max dim for data points. */


/* 
 * Function to get the dimensions of the matrix from the file 
 * Input: fileName - name of the file containing the matrix
 * Output: n - number of rows
 *         d - number of columns
 */
void getDimension(const char *fileName, int* n, int* d) {
    char line[MAX_ROW_LEN];
    char *token;
    FILE *file;

    *n = 0, *d = 0;

    file = fopen(fileName, "r");
    if (file == NULL) {
        printf("An Error Has Occurred");
        exit(1);
    }

    if (fgets(line, sizeof(line), file) != NULL) {
        token = strtok(line, " \t\n");
        while (token != NULL) {
            (*d)++;
            token = strtok(NULL, " \t\n");
        }
        (*n)++;
    }

    while (fgets(line, sizeof(line), file) != NULL) {
        (*n)++;
    }

    fclose(file);
}



/* 
 * Function to read data from a file and return it as a matrix 
 * Input: filename - name of the file containing the data
 *        n - number of rows
 *        d - number of columns
 * Return: Matrix - the data points from the file as a matrix
 */
Matrix readData(const char* filename, int n, int d) {
    char line[MAX_ROW_LEN];
    int row, col;
    char *token;
    FILE *file;
    Matrix X;

    X = createZeroMatrix(n, d);

    file = fopen(filename, "r");

    if (file == NULL) {
        printf("An Error Has Occurred");
        exit(1);
    }

    row = 0;

    /* Reading file line by line by tokenizing. */
    while (fgets(line, sizeof(line), file)) {
        token = strtok(line, " ");
        col = 0;

        while (token != NULL) {
            char *endptr;
            X.data[row][col] = strtod(token, &endptr);
            token = strtok(NULL, " ");
            col++;
        }
        row++;
    }

    fclose(file);

    return X;
}


/* 
 * Function to compute the similarity matrix 
 * Input: X - data matrix (n x d)
 * Return: Matrix - similarity matrix (n x n)
 */
Matrix sym(Matrix X){
    double *currentVector, *otherVector;
    int current, other;
    double distance;
    Matrix A;

    A = createZeroMatrix(X.rows, X.rows);

    for (current = 0; current < X.rows; current++){

        for (other = 0; other < X.rows; other++){

            if (current != other){
                currentVector = X.data[current];
                otherVector = X.data[other];

                distance = squaredEuclideanDistance(currentVector, otherVector, X.cols);
                A.data[current][other] =  exp((distance / -2));
            }
            else{
                A.data[current][other] = 0.0;
            }
        }
    }

    return A;
}


/* 
 * Function to compute the diagonal degree matrix 
 * Input: A - similarity matrix (n x n)
 * Return: Matrix - diagonal degree matrix (n x n)
 */
Matrix ddg(Matrix A){
    int diag;
    Matrix D;

    D = createZeroMatrix(A.rows, A.cols);

    for (diag = 0; diag < A.rows; diag++){
        D.data[diag][diag] = sumRow(A, diag);
    }

    return D;
}

/* 
 * Function to compute the normalized Laplacian matrix 
 * Input: D - diagonal degree matrix (n x n)
 *        A - similarity matrix (n x n)
 * Return: Matrix - normalized Laplacian matrix (n x n)
 */
Matrix norm(Matrix D, Matrix A){
    Matrix W, T;

    T = powerDiagMatrix(D, (-0.5));
    W = multiplyMatrix(T, A);
    W = multiplyMatrix(W, T);

    freeMatrix(T);

    return W;
}

/* 
 * Python wrapper function for different goals:
 * Input: goal - goal type:
            i. sym: Calculate and output the Similarity Matrix
            ii. ddg: Calculate and output the Diagonal Degree Matrix
            iii. norm: Calculate and output the Normalized Similarity Matrix
 *        fileName - name of the file containing the data
 * Return: Matrix - the result based on the specified goal
 */
Matrix symnmf(char *goal, char *fileName){
    int n, d;
    Matrix X;
    Matrix A;
    Matrix D;
    Matrix W;

    getDimension(fileName, &n, &d);

    X = readData(fileName, n, d);

    if (strcmp(goal,"sym") == 0){
        A = sym(X);
        freeMatrix(X);  
        return A;
    }
    if (strcmp(goal,"ddg") == 0){
        A = sym(X);
        D = ddg(A);
        freeMatrix(X);  
        freeMatrix(A);  
        return D;
    }
    if (strcmp(goal,"norm") == 0){
        A = sym(X);
        D = ddg(A);
        W = norm(D, A);
        freeMatrix(X);  
        freeMatrix(A);  
        freeMatrix(D);  
        return W;
    }

    freeMatrix(X);
    return X;
}


/* 
 * Python wrapper helper function to update H matrix for convarge_H
 * Input: H_current - current H matrix (n x k)
 *        W - weight matrix (n x n)
 * Return: Matrix - updated H matrix (n x k)
 */
Matrix update_H(Matrix H_current, Matrix W) {
    Matrix denominator = multiplyMatrix(multiplyMatrix(H_current, transposeMatrix(H_current)), H_current);
    Matrix H_new = createZeroMatrix(H_current.rows, H_current.cols);
    Matrix nominator = multiplyMatrix(W, H_current);
    double beta = 0.5;
    int i, j;

    for (i = 0; i < H_current.rows; i++) {
        for (j = 0; j < H_current.cols; j++) {
            H_new.data[i][j] = H_current.data[i][j] * (1 - beta + beta * (nominator.data[i][j] / denominator.data[i][j]));
        }
    }

    freeMatrix(nominator);
    freeMatrix(denominator);

    return H_new;
}


/* 
 * Python wrapper function to iteratively update H matrix until convergence 
 * Input: H - initial H matrix (n x k)
 *        W - weight matrix (n x n)
 *        eps - convergence threshold. def = 0.0001
 *        iter - maximum number of iterations. def = 300
 * Return: Matrix - converged H matrix (n x k)
 */
Matrix converge_H(Matrix H, Matrix W, double eps, int iter) {
    Matrix H_new = createZeroMatrix(H.rows, H.cols);
    int k;

    for (k = 0; k < iter; k++) {
        H_new = update_H(H, W);

        if (frobeniusNorm(H_new, H) < eps) {
            printf("Converged after %d iterations.\n", k + 1);
            break;
        }
        
        H = H_new;
    }

    return H_new;
}


/* 
 * Main function to run different goals based on input arguments 
 * Input: argc - number of command-line arguments
 *        argv - array of command-line arguments
 * Return: int - exit status
 */
int main(int argc, char *argv[]) {
    const char *fileName;
    char *goal;
    int n, d;
    Matrix X;
    Matrix A;
    Matrix D;
    Matrix W;

    if (argc > 0){
        goal = argv[1];
        fileName = argv[2];
    }
    else{
         printf("An Error Has Occurred\n");
         exit(1);
    }

    getDimension(fileName, &n, &d);

    X = readData(fileName, n, d);

    if ((strcmp(goal,"sym") == 0) || (strcmp(goal,"ddg") == 0) || (strcmp(goal,"norm") == 0)) {
        A = sym(X);

        if ((strcmp(goal,"ddg") == 0) || (strcmp(goal,"norm") == 0)){
            D = ddg(A);

            if (strcmp(goal,"norm") == 0){
                W = norm(D, A);
            }
        }
        

            if (strcmp(goal, "sym") == 0){
                    printMatrix(A);
                    freeMatrix(A);
            }
            else if (strcmp(goal, "ddg") == 0){
                    printMatrix(D);
                    freeMatrix(D);
                    freeMatrix(A); 
            } 
            else if (strcmp(goal, "norm") == 0){
                    printMatrix(W);
                    freeMatrix(W);
                    freeMatrix(D); 
                    freeMatrix(A); 
        }

    } 
    else{
        printf("An Error Has Occurred");
        exit(1);
    }

    return 0;
}

File: src/symnmf.py
import sys
import numpy as np
import mysymnmf as symnmf


def sys_arguments():
    """
    Parses command-line arguments.

    Expects three sys arguments:
    1. K (int): Number of required clusters.
    2. goal (str): Can get the following values:
        i. symnmf: Perform full the symNMF.
        ii. sym: Calculate and output the similarity matrix.
        iii. ddg: Calculate and output the Diagonal Degree Matrix.
        iv. norm: Calculate and output the normalized similarity matrix.
    3. file_name (str): The path to the Input file, it will contain N data points for all above
        goals, the file extension is .txt

    :return: k, goal, file_name
    """
    k, goal, file_name = None, None, None
    
    if len(sys.argv) != 4:
        raise ValueError(len(sys.argv), "An Error Has Occrred")

    try:
        k = int(sys.argv[1])
        goal = sys.argv[2]
        file_name = sys.argv[3]

        if goal not in ["symnmf", "sym", "ddg", "norm"]:
            raise ValueError("An Error Has Occrred")

    except ValueError:
        print("An Error Has Occrred")
        sys.exit(1)

    return k, goal, file_name


def read_data(file_name: str) -> np.ndarray:
    """
    Reads the data from the input file.
    :param file_name: The name of the text file containing the data.
    :return: The data stored in the file as a NumPy array.
    """
    with open(file_name, 'r') as f:
        lines = f.readlines()
        data = [[float(x) for x in line.strip().split(',')] for line in lines]
    return np.array(data)


def h_initialization(k: int, n: int, m: float) -> np.ndarray:
    """
    Randomly initialize H with values from the interval [0, 2 ∗ sqrt(m/k)].
    :param k: number of clusters
    :param n: number of data points
    :param m: the average of all entries of W.
    :return: lower dimension non-negative matrix H [n×k].
    """
    np.random.seed(0)
    np.random.uniform()

    upper_bound = 2 * np.sqrt(m / k)

    h = np.random.uniform(low=0, high=upper_bound, size=(n, k))

    return h


def update_H_until_convergence(H, W, epsilon=1e-5, max_iterations=100):
    """
    Update H using the given rule until convergence criteria are met.

    Args:
    - W: Matrix W (numpy array)
    - H: Initial matrix H (numpy array)
    - beta: Update parameter (default is 0.5)
    - max_iterations: Maximum number of iterations (default is 100)
    - epsilon: Convergence threshold (default is 1e-5)

    Returns:
    - Updated matrix H after convergence or reaching max iterations.
    """
    beta=0.5
    prev_H = H.copy()
    
    for t in range(max_iterations):
        WH = np.dot(W, H)
        HHTH = np.dot(np.dot(H, H.T), H)
        H = H * (1 - beta + beta * (WH / HHTH))

        # Check convergence
        diff_norm = np.linalg.norm(H - prev_H, 'fro') 
        if diff_norm < epsilon:
            print(f"Converged after {t+1} iterations.")
            break

        prev_H = H.copy()

    else:
        print(f"Warning: Max iterations ({max_iterations}) reached without convergence.")

    return H


def print_np_list(np_list):
    """
    Print np list as described in 2.1.4: 
    For example:
        0.0600,0.0100
        0.0100,0.0500
        0.0100,0.0400
        0.0200,0.0400
        0.0500,0.0200
    Args:
        np_list: numpy list
    """
    for row in np_list:
        print(",".join(f"{value:.4f}" for value in row))
        

def symNMF(x, k, n, epsilon=0.0001, max_iter=300):
    W = symnmf.symnmf_c('norm', x)
    H_init = h_initialization(k=k, n=n, m=np.mean(W))
    H_final = symnmf.converge_h_c(H_init, W, epsilon, max_iter)

    labels = np.argmax(H_final, axis=1)
    
    return labels
    
    
    
def main():
    np.random.seed(0)
    
    k, goal, file_name = sys_arguments()
    x = read_data(file_name=file_name)
    
    if (goal == "symnmf"):
        n, _ = x.shape  
        W = symnmf.symnmf_c('norm', x)
        H_init = h_initialization(k=k, n=n, m=np.mean(W))
        epsilon = 0.0001
        max_iter = 300
        H_final = symnmf.converge_h_c(H_init, W, epsilon, max_iter)

        print_np_list(H_final)
        
    elif(goal == "sym"):
        A = symnmf.symnmf_c('sym', x)
        
        print_np_list(A)
        
    elif(goal == "ddg"):
        D = symnmf.symnmf_c('ddg', x)
        
        print_np_list(D)

    elif(goal == "norm"):   
        W = symnmf.symnmf_c('norm', x)
        
        print_np_list(W)
        
        
if __name__ == "__main__":
    main()
    
    
    
    
    
    
